src/test/compile-fail/lint-no-drop-on-repr-extern.rs

```diff
@@ -0,0 +1,40 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Check we reject structs that mix a `Drop` impl with `#[repr(C)]`.
+//
+// As a special case, also check that we do not warn on such structs
+// if they also are declared with `#[unsafe_no_drop_flag]`
+
+#![feature(unsafe_no_drop_flag)]
+#![deny(drop_with_repr_extern)]
+
+#[repr(C)] struct A { x: Box<i8> }
+
+struct B { x: Box<i8> }
+
+#[repr(C)] struct C { x: Box<i8> }
+//~^ ERROR the `#[repr(C)]` attribute is attached here
+
+impl Drop for C { fn drop(&mut self) { } }
+//~^ ERROR implementing Drop adds hidden state to types, possibly conflicting with `#[repr(C)]`
+
+#[unsafe_no_drop_flag]
+#[repr(C)] struct D { x: Box<i8> }
+
+impl Drop for D { fn drop(&mut self) { } }
+
+fn main() {
+    let a = A { x: Box::new(3) };
+    let b = B { x: Box::new(3) };
+    let c = C { x: Box::new(3) };
+    let d = D { x: Box::new(3) };
+    println!("{:?}", (*a.x, *b.x, *c.x, *d.x));
+}
```

Comment:
Could you expand this test case to include an `enum` as well? You've also matched against `ty_closure` above, and I'm not sure how one can actually express that, but it might be nice to see a test case for that as well.